pydantic/generics.py

```diff
@@ -79,9 +83,28 @@ def __concrete_name__(cls: Type[Any], params: Tuple[Type[Any], ...]) -> str:
 
 
 def resolve_type_hint(type_: Any, typevars_map: Dict[Any, Any]) -> Type[Any]:
-    if hasattr(type_, '__origin__') and getattr(type_, '__parameters__', None):
-        concrete_type_args = tuple([typevars_map[x] for x in type_.__parameters__])
-        return type_[concrete_type_args]
+    type_args = _get_args(type_)
+    if type_args:
+        concrete_type_args = tuple(resolve_type_hint(arg, typevars_map) for arg in type_args)
+        origin_type = _get_origin(type_)
+        if origin_type is not None:
+            try:
+                return origin_type[concrete_type_args]
+            except TypeError:
+                pass
+            try:
+                # finally try to get origin directly from typing module
+                # in particular this can happen for python < 3.9 where the origin
+                # is a builtin type, see https://www.python.org/dev/peps/pep-0585
+                origin_type = getattr(typing, type_._name)
+                return origin_type[concrete_type_args]
+            except (TypeError, AttributeError):
+                pass
+        # this should not be reached
+        raise TypeError(f'Could not resolve type origin: {type_}')  # pragma: no cover
+
+    if lenient_issubclass(type_, GenericModel) and not type_.__concrete__:
+        return type_[tuple(resolve_type_hint(t, typevars_map) for t in type_.__parameters__)]
     return typevars_map.get(type_, type_)
```

Comment:
Found another edge case, related to `Callable`. It turns out tat that the "input arguments" part of the type is a list of types, which means that `resolve_type_hints` fails. Here's a test case:

```py
@skip_36
def test_generic_with_callable():
    T = TypeVar('T')
    U = TypeVar('U')

    class VeryInnerModel(GenericModel, Generic[T, U]):
        t: T
        u: U
        some_callable: Callable[[Optional[int], str], None]

    VeryInnerModel[str, U]
```

This results in the following failure: `TypeError: unhashable type: 'list'` when `[Optional[int], str]` is used as a key in the `typevars_map`

Reply:
Turns out there are a number of problems with `Callable`:
1. Need to handle it in `resolve_type_hint` properly
2. Need to handle it in `_has_typevar` properly